### PR TITLE
chore: add benchmarks

### PR DESCRIPTION
**Pull request by Ashad-h**
### Purpose

This PR add benchmarks for main read an write operations. #16 
This benchmark uses the [ANN_SIFT10K](http://corpus-texmex.irisa.fr/) dataset with 10 000 vectors of dimension 128.

### Approach

I added 2 dev dependencies:
- criterion.rs: a statistics-driven benchmark library
- byteorder: use to read the dataset

I added 3 benchmarks:
- Adding values in the database (10 000 vectors)
- Building the graph: 1 000 vectors (because it's very slow with 10 000 vectors. I think it's still relevant since the benchmarks average over 100 iterations)
- Searching the graph: search neighbours of 1 vector in a 10 000 vectors database

There may be other benchmarks to do.

You must download the dataset before running the benchmark:

```bash
curl ftp://ftp.irisa.fr/local/texmex/corpus/siftsmall.tar.gz --output data/siftsmall.tar.gz
tar xvf data/siftsmall.tar.gz
cargo bench
```

TODO:

- [ ] Find a more elegant way to download the dataset (within the "cargo bench" command)
- [ ] Add tests for other operations
- [ ] Add tests in real condition ( through the API)

### Testing

- [ ] I have tested this PR locally.
- [ ] I added tests to cover my changes, if not applicable, I have added a reason why.

How did you test this PR? Please provide a description of the tests that you ran to verify your changes.

How should the reviewer test this PR? If applicable, have you added tests to cover your changes?

### Chore checklist

- [ ] I have updated the documentation accordingly.
- [x] I have added comments to most of my code, particularly in hard-to-understand areas.

### CODE REVIEW

**Comment by Ashad-h** [comment]
#16

**Comment by Ashad-h** [comment]
Close #16

**Comment by Ashad-h** [comment]
#16

**Comment by edwinkys** [comment]
Hey, thanks for submitting the PR. I checked the PR locally and tested it. Everything is working well. We just need to format the code and everything should be good.

File: Cargo.toml [hunk 1]
```diff
@@ -27,3 +27,12 @@ serde_json = "1.0.108"
 lto = true
 opt-level = "z"
 codegen-units = 1
+
+[dev-dependencies]
+criterion = { version = "0.5.1", features = ["html_reports"] }
```
Comment (edwinkys) [inline Cargo.toml:32]: Can we separate this to its own section:

```toml
[dev-dependencies.criterion]
version = "0.5.1"
features = ["html_reports"]

[dev-dependencies]
byteorder = "1.5.0"
```

File: benches/benchmarks/graphs.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
+use std::collections::HashMap;
+use criterion::{criterion_group, Criterion};
+use oasysdb::db::database::*;
+use crate::benchmarks::utils;
+
+
+fn create_database(path: String, dimension: usize) -> Database {
```
Comment (edwinkys) [inline benches/benchmarks/graphs.rs:7]: Can we move this utility functions to either `benchmarks/mod.rs` along with the `add_values`, `build_graph`, and `search_graph` functions.

File: benches/benchmarks/graphs.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    db.create_graph(config).unwrap();
+}
+
+fn search_graph(db: &Database, name: &str, query: &Embedding, k: usize) {
+    // Search the graph
+    let _ = db.query_graph(name, query.clone(), k).unwrap();
+}
+
+
+fn bench_add_values(c: &mut Criterion) {
```
Comment (edwinkys) [inline benches/benchmarks/graphs.rs:36]: Can we move this benchmark function to its own module dedicated for values operation: `benchmarks/values.rs`.

File: benches/benchmarks/graphs.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    // Load the query dataset
+    let base = utils::load_base_dataset();
+    let dimension = base[0].len();
+
+    c.bench_function("add values (10_000 vector of dim 128)",
```
Comment (edwinkys) [inline benches/benchmarks/graphs.rs:41]: Is it possible to make this function calls less nested along with the other bench_functions call from the other benchmarks?

**Review by edwinkys — COMMENTED** [review]
The functionality is looking good. The code just need some formatting.

**Comment by edwinkys** [comment]
As per our discussion on Discord, I will close this PR since we pivot and rewrite the database as an embeddable vector database.

**Pull request closed without merging**